### PR TITLE
Refactor dependencies on dlmalloc

### PR DESCRIPTION
**Pull request by mjp41**
I have been looking into using different allocators inside the enclave.  There are a few direct dependencies on dlmalloc in the OE code base that makes this difficult. The changes in this PR contain the use to dlmalloc to a single compilation unit, and refactor the code to cope with this.

* The PR introduces new symbols oe_internal_malloc, oe_internal_free, etc which are directly provided by dlmalloc (but potentially other allocators) and are used inside of directly calling dlmalloc routines. 
* The debug allocator is factor to call the new symbols

* A separate compilation unit is introduced for dlmalloc, so that the oe_malloc/oe_free code is separate from the code for compiling dlmalloc. 

* The PR also introduces some thread local stubs for allocators that use thread local (TCS) state.

* The PR also removes the dlmalloc.h, as nothing should depend on this, and dlmalloc is a single file. Including this would be taking a specific dependency, so I have removed it to stop that potentially happening.

### CODE REVIEW

**Comment by mjp41** [comment]
CC @mikbras

File: enclave/core/sgx/atexit.c [hunk 1]
```diff
@@ -41,9 +41,10 @@ static oe_spinlock_t _spin = OE_SPINLOCK_INITIALIZER;
 
 static oe_atexit_entry_t* _new_atexit_entry(void (*func)(void*), void* arg)
 {
+    extern void* _oe_malloc(size_t);
     oe_atexit_entry_t* entry;
 
-    if ((entry = (oe_atexit_entry_t*)oe_sbrk(sizeof(oe_atexit_entry_t))) ==
+    if ((entry = (oe_atexit_entry_t*)_oe_malloc(sizeof(oe_atexit_entry_t))) ==
```
Comment (anakrish) [inline enclave/core/sgx/atexit.c]: Is this safe? Is there a guarantee that _oe_malloc will not rely on atexit/order?

File: 3rdparty/libunwind/stubs.h [hunk 2]
```diff
@@ -90,7 +90,7 @@ static __inline void* __libunwind_mmap(
     int fd,
     off_t offset)
 {
-    void* dlmemalign(size_t alignment, size_t size);
+    void* _oe_memalign(size_t alignment, size_t size);
```
Comment (anitagov) [inline 3rdparty/libunwind/stubs.h]: The underscore prefix i.e. _ should be dropped according to the OE coding standards as the single underscore prefix is reserved for local definitions. Please change this to oe_memalign.

File: 3rdparty/libunwind/stubs.h [hunk 3]
```diff
@@ -112,13 +112,13 @@ static __inline void* __libunwind_mmap(
 
 static __inline int __libunwind_munmap(void* addr, size_t length)
 {
-    extern void dlfree(void *ptr);
+    extern void _oe_free(void *ptr);
```
Comment (anitagov) [inline 3rdparty/libunwind/stubs.h]: This should be oe_free

File: enclave/core/sgx/dlmalloc/dlmalloc_wrapper.c [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+#pragma GCC diagnostic ignored "-Wparentheses-equality"
+#endif
+
+/* Indirection to alloc other allocators */
+#define dlmalloc         _oe_malloc
+#define dlcalloc         _oe_calloc
+#define dlrealloc        _oe_realloc
+#define dlmemalign       _oe_memalign
+#define dlposix_memalign _oe_posix_memalign
+#define dlfree           _oe_free
```
Comment (anitagov) [inline enclave/core/sgx/dlmalloc/dlmalloc_wrapper.c]: All these should have the leading underscore dropped as mentioned above.

File: enclave/core/sgx/internalmalloc.h [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+
+void* _oe_malloc(size_t);
+void  _oe_free(void*);
+void* _oe_calloc(size_t, size_t);
+void* _oe_realloc(void*, size_t);
+void* _oe_memalign(size_t alignment, size_t size);
+int   _oe_posix_memalign(void** memptr, size_t alignment, size_t size);
+
+void _oe_alloc_thread_teardown();
+void _oe_alloc_thread_startup();
```
Comment (anitagov) [inline enclave/core/sgx/internalmalloc.h]: Missing EOF character.

File: enclave/core/sgx/atexit.c (hunk 1 above)
Comment (mjp41) [reply enclave/core/sgx/atexit.c]: So the allocators I know about are very careful not to depend on anything to do with initialisation order.  jemalloc and dlmalloc both on every call check that they are initialised.  I don't know of any allocators that call atexit. Some call a pthread at exit during the per thread initialisation, but again have to be very careful about ensuring the allocator is sufficiently initialised, before making this call.

File: 3rdparty/libunwind/stubs.h (hunk 3 above)
Comment (sylvanc) [reply 3rdparty/libunwind/stubs.h]: As for the previous comment, this is not `oe_free`, but an internal symbol.

File: enclave/core/sgx/dlmalloc/dlmalloc_wrapper.c (hunk 4 above)
Comment (sylvanc) [reply enclave/core/sgx/dlmalloc/dlmalloc_wrapper.c]: As for the previous comment, these are separate symbols.

File: 3rdparty/libunwind/stubs.h (hunk 2 above)
Comment (sylvanc) [reply 3rdparty/libunwind/stubs.h]: Can someone from the OE team please provide guidance here? The headers in `include/openenclave/internal` use a single prefixed underscore in some places (structs), a double prefixed underscore in others (files.h, globals.h, etc), but the preference seems to be for no descriminator for internal but wider than module-scope functions. We're happy to use any name for these that suits the OE team, of course!

**Comment by anitagov** [comment]
@mikbras is reviewing this PR at length and will provide his feedback on the refactoring, naming conventions and more.

File: 3rdparty/libunwind/stubs.h (hunk 2 above)
Comment (mikbras) [reply 3rdparty/libunwind/stubs.h]: I'm sorry for the inconsistencies. Our coding convention is to use single underscores to introduce static function names. We didn't do a good job of publicizing this and some of our team members missed it. Of the options above, I liked "oe_internal_malloc()".

File: enclave/core/sgx/atexit.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (mjp41) [reply enclave/core/sgx/atexit.c]: Good catch.

**Comment by mjp41** [comment]
@mikbras I have switched to using oe_internal_*.

> struct called oe_allocator_t. 

I think this is going to introduce an additional layer of indirection on every malloc/free call, and unless you enable LTO this is not going to be optimised away.  

> Also, I'm wondering at what level we want the allocator to be pluggable. Should one be able to plug in a new allocator for application code but allow oecore to continue using its own allocator?

I am concerned at least with SGX the space is so constrained, the additional fragmentation from two allocators may be too much.

> should we consider weak aliases for symbols like malloc, realloc,

Maybe, I think how we correctly enable multiple allocators is interesting. Do you want the debug allocator in oe, actually to be linked in, or to wrap the underlying allocator if it is overridden. 

My main aim with this PR, was to put sufficient hooks and refactorings in place, so that changes to oe, won't affect my experiments for now, so I can track master without issue.

**Comment by mikbras** [comment]
I apologize for taking so long. I am working with your branch today and will try to complete my review so we can get this checked in by tomorrow. Thank you for your patience.

**Comment by sylvanc** [comment]
Thanks!

**Comment by mikbras** [comment]
Hi Matthew:

Building on your work, I created a new PR that uses weak-references to achieve the same goals. Please let me know what you think.

https://github.com/Microsoft/openenclave/pull/1420

Thanks,
Mike

**Comment by mikbras** [comment]
Hello Matthew:

Building on your PR, I took the concept a little further by factoring out dlmalloc int **liboealloc.a**, so now you can just replace this library as needed. Also, I made all the functions weakly-aliased so you can alternatively just override them by providing strongly-aliased alternatives.

Please see https://github.com/Microsoft/openenclave/pull/1420 for more details.

Thanks,
Mike

**Comment by johnkord** [comment]
This PR is superceded by @mikbras #1420

**Pull request closed without merging**